yt_dlp/downloader/niconico.py

```diff
@@ -94,47 +103,166 @@ def communicate_ws(reconnect):
                     elif data.get('type') == 'error':
                         self.write_debug(data)
                         message = try_get(data, lambda x: x['body']['code'], str) or recv
-                        return DownloadError(message)
+                        raise DownloadError(message)
                     elif self.ydl.params.get('verbose', False):
                         if len(recv) > 100:
                             recv = recv[:100] + '...'
                         self.to_screen('[debug] Server said: %s' % recv)
 
+        stopped = threading.Event()
+
         def ws_main():
             reconnect = False
-            while True:
+            while not stopped.is_set():
                 try:
-                    ret = communicate_ws(reconnect)
-                    if ret is True:
-                        return
-                except BaseException as e:
-                    self.to_screen('[%s] %s: Connection error occured, reconnecting after 10 seconds: %s' % ('niconico:live', video_id, str_or_none(e)))
-                    time.sleep(10)
-                    continue
-                finally:
+                    communicate_ws(reconnect)
+                    break  # Disconnected
+                except BaseException as e:  # Including TransportError
+                    if stopped.is_set():
+                        break
+
+                    self.m3u8_lock.clear()  # m3u8 url may be changed
+
+                    self.to_screen('[%s] %s: Connection error occured, reconnecting after %d seconds: %s' % ('niconico:live', video_id, self._WEBSOCKET_RECONNECT_DELAY, str_or_none(e)))
+                    time.sleep(self._WEBSOCKET_RECONNECT_DELAY)
+
                     reconnect = True
 
+            self.m3u8_lock.set()  # Release possible locks
+
         thread = threading.Thread(target=ws_main, daemon=True)
         thread.start()
 
-        return dl.download(filename, new_info_dict)
+        try:
+            yield self
+        finally:
+            stopped.set()
+            self.ws.close()
+            thread.join()
+
+    def _master_m3u8_url(self):
+        """ Get the refreshed manifest url after WebSocket reconnection to prevent HTTP 403 """
+
+        self.m3u8_lock.wait()
+        return self.m3u8_url
+
+
+class NiconicoLiveFD(NiconicoLiveBaseFD):
+    """ Downloads niconico live without being stopped """
+
+    def real_download(self, filename, info_dict):
+        with self._ws_context(info_dict):
+            new_info_dict = info_dict.copy()
+            new_info_dict.update({
+                'protocol': 'm3u8',
+            })
+
+            return FFmpegFD(self.ydl, self.params or {}).download(filename, new_info_dict)
+
+
+class NiconicoLiveTimeshiftFD(NiconicoLiveBaseFD, HlsFD):
+    """ Downloads niconico live timeshift VOD """
+
+    _PER_FRAGMENT_DOWNLOAD_RATIO = 0.1
+
+    def real_download(self, filename, info_dict):
+        with self._ws_context(info_dict) as ws_context:
```

Comment:
I want to avoid adding more "protocols". Can't we keep the single protocol and do something like:
```suggestion
class NiconicoLiveFD(NiconicoLiveBaseFD):
    """Downloads niconico live/timeshift VOD"""

    _PER_FRAGMENT_DOWNLOAD_RATIO = 0.1

    def real_download(self, filename, info_dict):
        with self._ws_context(info_dict) as ws_context:
            if info_dict.get('is_live'):
                info_dict = info_dict.copy()
                info_dict['protocol'] = 'm3u8'
                return FFmpegFD(self.ydl, self.params or {}).download(filename, new_info_dict)

```